src/middlewared/middlewared/plugins/zfs.py

```diff
@@ -326,6 +327,141 @@ def promote(self, name):
+        zfs = libzfs.ZFS()
+        tasks_to_execute = []
+
+        date_now = datetime.datetime.now()
+        curtime = datetime.time(date_now.hour, date_now.minute)
+
+        for task in snap_tasks:
+            if task['task_last_run']:
+                snaptask_time_delta = task['task_last_run'] + datetime.timedelta(minutes=task['task_interval'])
+            else:
+                snaptask_time_delta = date_now
+
+            if curtime < task['task_begin'] or curtime > task['task_end'] or date_now < snaptask_time_delta or str(date_now.weekday()+1) not in task['task_byweekday']:
+                continue
+
+            try:
+                zfs.get_dataset(task['task_filesystem'])
+            except libzfs.ZFSException as err:
+                self.logger.error(f'{err} - Periodic snapshot task omitted.')
+            tasks_to_execute.append(task)
+
+        for task in tasks_to_execute:
+            task_name = task['task_name'] if task['task_name'] else f"task_{task['id']}"
+            self.middleware.call_sync(
+                'zfs.snapshot.do_create',
+                {'dataset': task['task_filesystem'],
+                'recursive': task['task_recursive'],
+                'name': f'autosnap_new_{task_name}_{date_now.strftime("%Y%m%d.%H%M")}',
+                'properties': {'autosnap:name': task_name, 'autosnap:retention': f'{task["task_ret_count"]}:{task["task_ret_unit"]}'}}
+            )
+
+            self.middleware.call_sync(
+                'zfs.snapshot_task.do_update',
+                task['id'],
+                {'task_last_run': f'{date_now.year}-{date_now.month}-{date_now.day} {date_now.hour}:{date_now.minute}'}
+            )
+
+            try:
+                repl_task = self.middleware.call_sync('replication.task.query', [('repl_snap_task', '=', task['id'])], {'get': True})
+            except IndexError:
+                pass
+            else:
+                self.middleware.call_sync(f"replication.task.start_{repl_task['repl_transport']}_replication", repl_task['id'])
+
+
+
+
+    @periodic(300)
+    async def autosnap_remove(self):
+
+        def calculate_snap_time_delta(date, task_ret_count, ret_unit):
+            if ret_unit == 'hour':
+                snaptask_time_delta = date + datetime.timedelta(hours=int(task_ret_count))
+            elif ret_unit == 'day':
+                snaptask_time_delta = date + datetime.timedelta(days=int(task_ret_count))
+            elif ret_unit == 'week':
+                snaptask_time_delta = date + datetime.timedelta(days=7 * int(task_ret_count))
+            elif ret_unit == 'month':
+                snaptask_time_delta = date + datetime.timedelta(days=int(30.436875 * task_ret_count))
+            elif ret_unit == 'year':
+                snaptask_time_delta = date + datetime.timedelta(days=int(365.2425 * task_ret_count))
+
+            return snaptask_time_delta
+
+        snaps = await self.middleware.call('zfs.snapshot.query')
```

Comment:
It seems more reasonable to use zfs subprocess call here for now since we don't do caching and libzfs is an order of magnitude slower than zfs subprocess if all we need here is the `autosnap:retention` property.